src/compiler/mir/ops/traverser.rs

```diff
@@ -30,9 +30,19 @@ impl<'a> ProgramTraverser<'a> {
     ) {
         debug!("Applying given Transformer to MIR");
 
+        // Declare every structure, but do not define the structures yet.
+        // This solves the problem of structures with reference cycles
+        debug!("Declare any structures");
+        for (id, ty) in self.mir.type_iter() {
+            match ty {
+                MirTypeDef::Structure { path, .. } => xfmr.declare_struct(id, path).unwrap(),
+                _ => (),
+            }
+        }
```

Comment:
This could be moved into the recursive block in the `map_type`, for when a structure is encountered.